crates/hir-def/src/item_tree.rs

```diff
@@ -696,9 +696,33 @@ pub struct TraitAlias {
     pub ast_id: FileAstId<ast::TraitAlias>,
 }
 
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub struct GenericParamGroups {
+    // pub target_trait: Option<Vec<String>>,
+    pub self_ty: Option<Vec<String>>,
+}
+
+impl GenericParamGroups {
+    fn from_impl(impl_: &ast::Impl) -> Self {
+        Self {
+            // target_trait: Self::get_arg_list(impl_.trait_()),
+            self_ty: Self::get_arg_list(impl_.self_ty()),
+        }
+    }
+
+    fn get_arg_list(ty: Option<ast::Type>) -> Option<Vec<String>> {
+        let list = match ty? {
+            ast::Type::PathType(p) => p.path()?.segment()?.generic_arg_list(),
+            _ => None,
+        };
+        Some(list?.generic_args().map(|arg| arg.syntax().text().to_string()).collect())
+    }
+}
+
```

Comment:
This is the wrong way too approach this. The fix to the issue here should not involve touching item tree whatsoever.

Reply:
I has to touch it, I'm afraid.

Let's have a look on few relevant lines from the test:
```
struct S(Foo<i32, bool>);
...
impl<A, B, C> Trait<A, B> for Foo<C, B> {
```
In this case the assist should turn `C` and `B` into `i32` and `bool` repectively. In general, it should get `Foo`-related generic arguments from the trait implementation and correlate them `Foo`'s parameters within the structure. Consequently, we have to memorize the "distribution" of the generic arguments in `impl` lines; that's exactly what happens in this file - the names of args pertaining to `self_ty` are being saved.

Reply:
That data is already encoded in the `sekf_ty` `TypeRef` in the item tree. So you just need to fetch that from there instead, that is check if the `TypeRef` is a `TypeRef::Path` and take the generic args from that path instead

Reply:
Thank you! Now these data are fetched from `Type` instead of `TypeRef`, but anyway the item tree is unchanged